monai/metrics/panoptic_quality.py

```diff
@@ -0,0 +1,282 @@
+    iou_sum = paired_iou.sum()
+
+    if output_confusion_matrix:
+        return torch.as_tensor([tp, fp, fn, iou_sum], device=pred.device)
+
+    metric_name = _check_panoptic_metric_name(metric_name)
+    if metric_name == "rq":
+        return torch.as_tensor(tp / (tp + 0.5 * fp + 0.5 * fn + smooth_numerator), device=pred.device)
+    if metric_name == "sq":
+        return torch.as_tensor(iou_sum / (tp + smooth_numerator), device=pred.device)
+    return torch.as_tensor(iou_sum / (tp + 0.5 * fp + 0.5 * fn + smooth_numerator), device=pred.device)
+
+
+def _get_id_list(gt: torch.Tensor):
+    id_list = list(gt.unique())
+    # ensure id 0 is included
+    if 0 not in id_list:
+        id_list.insert(0, torch.tensor(0).int())
+
+    return id_list
+
+
+def _get_pairwise_iou(pred: torch.Tensor, gt: torch.Tensor, device: Union[str, torch.device] = "cpu"):
+    pred_id_list = _get_id_list(pred)
+    true_id_list = _get_id_list(gt)
+
+    pairwise_iou = torch.zeros([len(true_id_list) - 1, len(pred_id_list) - 1], dtype=torch.float, device=device)
+    true_masks: List[torch.Tensor] = []
+    pred_masks: List[torch.Tensor] = []
+
+    for t in true_id_list[1:]:
+        t_mask = torch.as_tensor(gt == t, device=device).int()
+        true_masks.append(t_mask)
+
+    for p in pred_id_list[1:]:
+        p_mask = torch.as_tensor(pred == p, device=device).int()
+        pred_masks.append(p_mask)
+
+    for true_id in range(1, len(true_id_list)):
+        t_mask = true_masks[true_id - 1]
+        pred_true_overlap = pred[t_mask > 0]
+        pred_true_overlap_id = list(pred_true_overlap.unique())
+        for pred_id in pred_true_overlap_id:
+            if pred_id == 0:
+                continue
+            p_mask = pred_masks[pred_id - 1]
+            total = (t_mask + p_mask).sum()
+            inter = (t_mask * p_mask).sum()
+            iou = inter / (total - inter)
+            pairwise_iou[true_id - 1, pred_id - 1] = iou
+
+    return pairwise_iou, true_id_list, pred_id_list
+
+
+def _get_paired_iou(pairwise_iou: torch.Tensor, match_iou: float = 0.5, device: Union[str, torch.device] = "cpu"):
+    if match_iou >= 0.5:
+        paired_iou = pairwise_iou[pairwise_iou > match_iou]
+        pairwise_iou[pairwise_iou <= match_iou] = 0.0
+        paired_true, paired_pred = torch.nonzero(pairwise_iou)[:, 0], torch.nonzero(pairwise_iou)[:, 1]
+        paired_iou = pairwise_iou[paired_true, paired_pred]
+        paired_true += 1
+        paired_pred += 1
+
+        return paired_iou, paired_true, paired_pred
+
+    pairwise_iou = pairwise_iou.cpu().numpy()
+    paired_true, paired_pred = linear_sum_assignment(-pairwise_iou)
+    paired_iou = pairwise_iou[paired_true, paired_pred]
+    paired_true = torch.as_tensor(list(paired_true[paired_iou > match_iou] + 1), device=device)
+    paired_pred = torch.as_tensor(list(paired_pred[paired_iou > match_iou] + 1), device=device)
+    paired_iou = paired_iou[paired_iou > match_iou]
```

Comment:
Shouldn't they be first filtered for `paired_iou > match_iou` and then use `linear_sum_assignment`?
I haven't mathematically worked it out but it seems that the values less than the threshold value (e.g., 0.4) add to the cost while it doesn't matter what value they have (e.g., it does matter it is 0.01 or 0.03 because they won't be a match any way). Any thought?

Reply:
Hi @drbeh , I get your point here, but I'm not sure if "filtering first, then mapping" is equivalent to "mapping first, then filtering". Therefore, I suggest here we follow the order of the original implementation (https://github.com/TissueImageAnalytics/CoNIC/blob/main/metrics/stats_utils.py#L143), what do you think?